tests/test_minio_repo.py

```diff
@@ -0,0 +1,149 @@
+from datetime import datetime
+from io import BytesIO
+from unittest import TestCase, mock
+
+import pytest
+from freezegun import freeze_time
+
+from websub.repos import Pattern, Subscription, SubscriptionsRepo, Id
+
+
+class MinioPatternTest(TestCase):
+    def test_to_key__when_empty__should_return_error(self):
+        with pytest.raises(ValueError):
+            Pattern(predicate='').to_key()
+
+    def test_to_key__when_contains_slashes__should_return_error(self):
+        with pytest.raises(ValueError):
+            Pattern(predicate='aa/bb').to_key()
+
+    def test_to_key__when_wildcard_without_dot__should_return_error(self):
+        with pytest.raises(ValueError):
+            Pattern(predicate='aa.bb*').to_key()
+
+    def test_to_key__when_predicate_valid__should_return_key(self):
+        assert Pattern('aaaa.bbbb.cccc').to_key() == "AAAA/BBBB/CCCC/"
+
+    def test_to_key__with_wildcard_in_predicate__should_be_handled(self):
+        assert Pattern('aaaa.bbbb.cccc.*').to_key() == "AAAA/BBBB/CCCC/"
+        assert Pattern('aaaa.bbbb.cccc.').to_key() == "AAAA/BBBB/CCCC/"
+
+    def test_to_layers__should_return_list_of_layers(self):
+        assert Pattern('aaaa.bbbb.cccc.*').to_layers() == [
+            'AAAA/',
+            'AAAA/BBBB/',
+            'AAAA/BBBB/CCCC/'
+        ]
+
+
+@freeze_time("2020-05-12 12:00:01")
+class SubscriptionTest(TestCase):
+    def setUp(self):
+        body = mock.MagicMock()
+        body.read.return_value = b'{"e": "2020-05-12 14:00:01", "c": "http://callback.com/1"}'
+
+        self.obj = {'Body': body}
+
+    def test_subscription_for_valid_object__should_have_callback_url(self):
+        subscription = Subscription(self.obj, 'some_name', now=datetime.utcnow())
+        assert subscription.is_valid
+        assert not subscription.is_expired
+        assert subscription.callback_url == 'http://callback.com/1'
+
+    @freeze_time("2020-05-12 15:00:01")
+    def test_subscription_for_expired_object__should_be_not_valid(self):
+        subscription = Subscription(self.obj, 'some_name', now=datetime.utcnow())
+        assert not subscription.is_valid
+        assert subscription.is_expired
+
+    def test_subscription__when_missing_callback__should_be_not_valid(self):
+        self.obj['Body'].read.return_value = b'{}'
+        subscription = Subscription(self.obj, 'some_name', now=datetime.utcnow())
+        assert not subscription.is_valid
+        assert subscription.error == "data missing required key:'c'"
+
+    def test_subscription__when_missing_expiration__should_be_valid(self):
```

Comment:
I made expiration key optional since it is so in WebSub.
@adamkerz are we fine with that?

Reply:
Yep, we definitely need a default though.